### PR TITLE
1584 sidenav

### LINKED ISSUE #1584: [SideNav] item styling/indentation is invalid
### What package is this for?

- [x] React
- [ ] Angular

### Describe the bug

Regression of SideNav styles with latest Carbon update. 

image

### LINKED ISSUE #1618: [SideNav] implementation discrepancies
### What package is this for?

- [x] React
- [ ] Angular

### Describe the bug

We noticed the following discrepancies in the SideNav component implementation. See the attached screenshots compared to what we designed and a list describing the issues.
 
SideNav - issues

1. There is an extra spacing above the first item in the implementation
2. The icon size is implemented as 16x16px but by design it should be 20x20px 
3. The text in the SideNav items shouldn’t be bold _(it should also be white, not sure why in the React implementation some items appear with grey text)_.
4. Indentation of child items with children doesn't match indent of child items with no additional children _(this is not clearly captured in the above screenshots comparison, but persists as an issue)_
5. The selected SideNav item should be shown with light background and left blue border _(this shows correctly on the React code but not in the TRIRIGA implementation)_
6. The height for child items should be 32px height instead of 48px

### Additional context

Storybook SideNav - component
https://carbon-addons-iot-react.com/?path=/story/watson-iot-sidenav--sidenav-component

AI Application Design SideNav - usage
https://pages.github.ibm.com/ai-applications/design/components/ui-shell-side-nav/usage

TRIRIGA SideNav - documentation
https://ibm.invisionapp.com/share/ZMNZSMJVSUY#/319510824_Navigation_Components_-_Side_Navigation

### PR DESCRIPTION
**Pull request by davidicus**
Closes #1584  and closes #1618

**Summary**
Refactored SideNav styles to fix regressions and more closely rely on base carbon styles

**Change List (commits, features, bugs, etc)**

- created _side-nav-carbon.scss to allow for our pattern modifications. We have both buttons and anchors right now. Carbon scopes their styles to the `a` element. This file allows for any element with the same class name as well as fix some default styling that are applied to buttons. 
- _side-nav.scss has been refactored to only include theme styles to allow for our dark sidenav pattern. 
- created _side-nav-multi-tier.scss to allow for 3 level nesting. FYI @ScottHale. This may help with #1606 
- SideNav.jsx I removed the large attribute and added the isRail attribute which bring in specific Carbon styles. 

**Acceptance Test (how to verify the PR)**

- Go to sideNav story and make sure sideNav items look as outlined in spec below

### CODE REVIEW

**Review by tay1orjones — APPROVED** [review]

**Comment by dianatran18** [comment]
Looks good, with a few minor fixes:
- When an L1 is expanded, there should be a lighter background color that stays persistent on the whole group. Right now it's only showing on hover
- When you expand out **Member** it auto selects **Link 3** (seems like a bug), but the selection shouldn't be made until the user clicks on it
- On the L2 selected state (the Member > Link 3 example) the text should stay regular, not semi-bold.

I spoke with Luke about the font weights, and we concluded that for L1s they should always be semi-bold, but for any L2s and L3s they should always be regular, selected or not. That aligns with what Carbon has in their UI shell kit too.

Example below shows what all of the changes above should look like

Screen Shot 2020-10-19 at 2 11 51 PM

I'll make an update to the design to make the L1s semi-bold as well. 

Let me know if you have any questions @davidicus !

File: src/components/SideNav/_side-nav-carbon.scss
Comment (bjornalm) [inline src/components/SideNav/_side-nav-carbon.scss:17]: her

File: src/components/SideNav/_side-nav-carbon.scss
Comment (bjornalm) [inline src/components/SideNav/_side-nav-carbon.scss:20]: Is there a variable or calculation for 65.99em? Seems oddly specific. At least we can put it in a variable here since it is used multiple times.

File: src/components/SideNav/_side-nav-carbon.scss
Comment (bjornalm) [inline src/components/SideNav/_side-nav-carbon.scss:51]: TODO left in code

File: src/components/SideNav/_side-nav-multi-tier.scss
Comment (bjornalm) [inline src/components/SideNav/_side-nav-multi-tier.scss:1]: Are there no carbon variables for these colors?

**Review by bjornalm — APPROVED** [review]
Looks good, only some minor comments/questions.

File: src/components/SideNav/_side-nav-multi-tier.scss
Comment (dianatran18) [reply src/components/SideNav/_side-nav-multi-tier.scss:1]: #f3f3f3 sounds like the old Gray 10 color, which is now #f4f4f4.

The equivalent color token name for the text in the Gray 10 theme is $inverse-01 though I believe (that's what I have on my spec)

**Review by tay1orjones — CHANGES_REQUESTED** [review]
@davidicus I'm placing a `changes requested` on here just so we don't accidentally merge this without addressing feedback from @dianatran18

File: src/components/SideNav/_side-nav-carbon.scss
Comment (davidicus) [reply src/components/SideNav/_side-nav-carbon.scss:20]: It is the opposite of the variables breakpoints that Carbon has that are min-width. Applied here so I am not having to duplicate more styles. But I will make a variable

**Comment by davidicus** [comment]
> Looks good, with a few minor fixes:
> 
> * When an L1 is expanded, there should be a lighter background color that stays persistent on the whole group. Right now it's only showing on hover
> * When you expand out **Member** it auto selects **Link 3** (seems like a bug), but the selection shouldn't be made until the user clicks on it
> * On the L2 selected state (the Member > Link 3 example) the text should stay regular, not semi-bold.
> 
> I spoke with Luke about the font weights, and we concluded that for L1s they should always be semi-bold, but for any L2s and L3s they should always be regular, selected or not. That aligns with what Carbon has in their UI shell kit too.
> 
> Example below shows what all of the changes above should look like
> 
> Screen Shot 2020-10-19 at 2 11 51 PM
> 
> I'll make an update to the design to make the L1s semi-bold as well.
> 
> Let me know if you have any questions @davidicus !

Thanks @dianatran18. I will make changes to 1 and 3. 2 is only selected because the story has that as the active link.

File: src/components/SideNav/_side-nav-multi-tier.scss
Comment (davidicus) [reply src/components/SideNav/_side-nav-multi-tier.scss:1]: yes I have updated textColor to be the `$inverse-01`

File: src/components/SideNav/_side-nav-carbon.scss
Comment (davidicus) [reply src/components/SideNav/_side-nav-carbon.scss:51]: lifted from carbon but I removed.

**Comment by davidicus** [comment]
@tay1orjones I have pushed up changes that Diana requested. There is a question about the selected font-weight as their code is different than the design kit but we are currently aligned with what Carbon has in code for that. We can merge this PR and if needed open up issue to change the font weight or we can wait for confirmation.

**Review by tay1orjones — APPROVED** [review]

**Comment by davidicus** [comment]
image

Added padding change to align header title with sidnav issues to address a carbon bug.

**Pull request merged**
